src/audio/copier.c

```diff
@@ -202,38 +208,97 @@ static struct comp_dev *create_host(struct comp_dev *parent_dev, struct copier_d
 	return dev;
 }
 
+static int init_dai(struct comp_dev *parent_dev, struct copier_data *cd,
+		    const struct comp_driver *drv,
+		    struct comp_ipc_config *config,
+		    struct ipc4_copier_module_cfg *copier,
+		    struct pipeline *pipeline,
+		    struct ipc_config_dai *dai,
+		    enum ipc4_gateway_type type)
+{
+	struct comp_dev *dev;
+	int ret;
+
+	create_endpoint_buffer(parent_dev, cd, config, copier);
+
+	dev = drv->ops.create(drv, config, dai);
+	if (!dev)
+		return -EINVAL;
+
+	if (dai->direction == SOF_IPC_STREAM_PLAYBACK)
+		pipeline->sink_comp = dev;
+	else
+		pipeline->source_comp = dev;
+
+	pipeline->sched_id = config->id;
+
+	list_init(&dev->bsource_list);
+	list_init(&dev->bsink_list);
+
+	ret = comp_dai_config(dev, dai, copier);
+	if (ret < 0)
+		return ret;
+
+	if (dai->direction == SOF_IPC_STREAM_PLAYBACK) {
+		comp_buffer_connect(dev, config->core, cd->endpoint_buffer[cd->endpoint_num],
+				    PPL_CONN_DIR_BUFFER_TO_COMP);
+		cd->bsource_buffer = true;
+	} else {
+		comp_buffer_connect(dev, config->core, cd->endpoint_buffer[cd->endpoint_num],
+				    PPL_CONN_DIR_COMP_TO_BUFFER);
+		cd->bsource_buffer = false;
+	}
+
+	cd->converter[0] = get_converter_func(&copier->base.audio_fmt, &copier->out_fmt, type,
+					      IPC4_DIRECTION(dai->direction));
+	if (!cd->converter[0]) {
+		comp_err(parent_dev, "failed to get converter type %d, dir %d",
+			 type, dai->direction);
+		return -EINVAL;
+	}
+
+	cd->endpoint[cd->endpoint_num] = dev;
+	cd->endpoint_num++;
+
+	return 0;
+}
+
 /* if copier is linked to non-host gateway, it will manage link dma,
  * ssp, dmic or alh. Sof dai component can support this case so copier
  * reuses dai component to support non-host gateway.
  */
-static struct comp_dev *create_dai(struct comp_dev *parent_dev, struct copier_data *cd,
-				   struct comp_ipc_config *config,
-				   struct ipc4_copier_module_cfg *copier,
-				   union ipc4_connector_node_id *node_id,
-				   struct pipeline *pipeline)
+static int create_dai(struct comp_dev *parent_dev, struct copier_data *cd,
+		      struct comp_ipc_config *config,
+		      struct ipc4_copier_module_cfg *copier,
+		      struct pipeline *pipeline)
 {
 	struct sof_uuid id = {0xc2b00d27, 0xffbc, 0x4150, {0xa5, 0x1a, 0x24,
 				0x5c, 0x79, 0xc5, 0xe5, 0x4b}};
+	int dai_index[IPC4_MAX_MULTIPLE_GTW_COUNT];
+	struct sof_alh_configuration_blob *alh_blob;
+	union ipc4_connector_node_id node_id;
 	enum ipc4_gateway_type type;
 	const struct comp_driver *drv;
 	struct ipc_config_dai dai;
-	struct comp_dev *dev;
+	int dai_count;
 	int ret;
+	int i;
 
 	drv = ipc4_get_drv((uint8_t *)&id);
 	if (!drv)
-		return NULL;
+		return -EINVAL;
 
 	parent_dev->ipc_config.type = SOF_COMP_DAI;
 	config->type = SOF_COMP_DAI;
-	create_endpoint_buffer(parent_dev, cd, config, copier);
 
 	memset(&dai, 0, sizeof(dai));
-	dai.dai_index = node_id->f.v_index;
-	dai.direction = node_id->f.dma_type % 2;
+	dai_index[0] = node_id.f.v_index;
```

Comment:
again why index 0, does it have a special meaning ? if so we should have a macro for this meaning so its easier to follow.